### ISSUE TITLE
Issues with DO on a simple toy model (ticket 1 of some)

### ISSUE THREAD

**Issue by max-privato** [OpenModelica/OpenModelica#7969]
### Description
I'm trying to use the Dynamic Optimisation feature of OpenModelica.

In doing so I came across a few issues, that make it very difficult to use it. Nevertheless, I'm convinced that the feature in principle is fantastic, and once it is made manageable it will enhance OM considerably (and Dymola has not it). Therefore I'll open a few tickets on the issues I found on this toy model.

**I hope there is interest from OSMC on these tickets. My university is interested, and if deemed feasible by OSMC could propose a Directly Funded Development on them.**

The "toy model" proposed here is intended to evaluate the optimal strategy to manage a system having some load power, an Internal-combustion-engine-based electricity generator, and a storage system. 
This may represent some stationary hybrid generation or a hybrid vehicle. 
Purpose of the optimisation is to find the optimal management of the storage so that to minimise fuel cost.
This is well represented by the following Modelica diagram:
![Picture1](https://user-images.githubusercontent.com/37046832/136050769-e4fc92bc-b239-4fa8-b962-690d78fe98ff.png)

The good news it that OM Dynamic Optimisation fonverges and converfes fast.

Below the comparison between the the expected optimal behaviour (right, model "ManualO"), and the one given by OM (left, model "OM_DO").

![Picture3](https://user-images.githubusercontent.com/37046832/136072340-493d0c75-8f52-4c69-9d58-0962e75e9d7f.png)

The OM result is only slightly suboptimal, implying a litltle larger cost than the optimal. mainly due to the fact that the cure outBatPow is not exactly zero where it should be, and icePow not exactly constant when outBatPow is positive in the upper-right diagram. 

However, this minor lack of precision is not a real issue. This example is very promising for DO in OM. 
A larger issue of this simulation is due to the fact that the model requires outBatPower to be limited not to overcome 20 kW:
```
  parameter Modelica.SIunits.Power maxBatPower = 20e3 "W";
  input Real outBatPow (min = -maxBatPower, max = maxBatPower);

```
and this limit is not taken into consideration by OM which lets outBatPow to reach 31.5 kW. Indeed OM by some reason does not accept limits on iput stated this way. The output message says: 
```
Optimizer Variables
========================================================
State[0]:storage.energy(start = 1.8e+07, nominal = 3.6e+07, min = 0, max = 3.6e+07, init = 1.8e+07)
State[1]:toGrams.y(start = 0, nominal = 1, min = -Inf, max = +Inf, init = 0)
Input[2]:outBatPow(start = 0, nominal = 1, min = -Inf, max = +Inf)

Finally I must say that this ticket's model has a limitation on icePowerConstr, which is apparently taken into consideration by the software, but not reported in the mesage browser:

```
  Real icePowerConstr(min = 0) = ipcp * toIcePow.y "ICE power contraint" annotation(
    isConstraint = true);

```
```

NOTE that this toy model will be used also for a few other tickets which I'll create very soon.

### Steps to Reproduce
see the above description; use the following file:
[ToyModelTicket1.zip](https://github.com/OpenModelica/OpenModelica/files/7288262/ToyModelTicket1.zip)

### Expected Behavior

1) The DO algorithm should apply the box limit to outBatPower.
2) In addition (this is a relatively minor issue) it should find the optimum more precisely. The check made above refers to the case in which no limitation on outBatPow is present
3) It should use the given limits on input variable instead of converting to inf
4) it is not clear why the software takes into consideration the given limit on icePowerConstr, but does not report this in the output message

### Version and OS
OM 1.18.0 WIn 10 64 bit

**Comment by max-privato** [comment]
aha, well spotted!
Thank you Vitalij for sorting this out. 
In addition to the missing limits on outBatPower, the ticket has something more to say: my point 2 of the list in the ticket description. Even though this is a limited issue, it seems strange that the expected solution is not found with larger precision.

**Comment by casella** [comment]
@max-privato should we improve something in the documentation, GUI or error messages to avoid such situations in the future?

**Comment by vruge** [comment]
it seems the inputs are excluded from the processing [here](https://github.com/OpenModelica/OpenModelica/blob/a1b18e11a79b12af0775042cf92b6c8a86da3175/OMCompiler/Compiler/SimCode/SimCodeUtil.mo#L7086), because the inputs are part of the know variables.

**Comment by vruge** [comment]
We get now

```bash
Optimizer Variables                                                                                                                                                                                                                          ========================================================                                                                                                                                                                                     State[0]:storage.energy(start = 1.8e+07, nominal = 3.6e+07, min = 0, max = 3.6e+07, init = 1.8e+07)                                                                                                                                          State[1]:toGrams.y(start = 0, nominal = 1, min = -Inf, max = +Inf, init = 0)                                                                                                                                                                 Input[2]:outBatPow(start = 0, nominal = 25000, min = -25000, max = 20000)                                                                                                                                                                    --------------------------------------------------------                                                                                                                                                                                     number of nonlinear constraints: 1    
```

// I relaxe the limit `outBatPow.min`. Otherwise the alogorithmen stuck in local infeasibility point.

fixed with #8030

**Comment by max-privato** [comment]
> fixed with #8030

good!
I'll check tomorrow.

**Comment by max-privato** [comment]
ok, I checked this ticket's models, and others deriving from them, on OM 1.19.0-dev-336.

Now the given limits on the input power are taken into account, but the resulting model does not converge.
The following package contains updated motels: EConst, without battery limits. and EConstLimPbat, with battery limits:

[ToyModelTicket1Upd.txt](https://github.com/OpenModelica/OpenModelica/files/7453856/ToyModelTicket1Upd.txt)

Do we have to accept that this model is now too complex to have an automatically computed optimal solution, or there is something more we can do to solve it?

**Comment by max-privato** [comment]
I checked #8030 with good results. 
This is real advancement!

Looking back at this ticket, I see that, except for the fact that input limits were not taken into account, the issues I found came from my inability to manage a few characteristics of this "no-optimica Dynamical optimisation" (NOODO).

Summarising how to turn my faulty model into a good one, one has to 
1. use a much lower tolerance (initially I used 1e-4; it resulted that at least 1e-7 is needed). This is not sufficient: if you do just this, no optimal solution is found
2. use continuous derivative on table toGperkWh => now (using also 1e-7 as tolerance) an "optimal solution" is found, which is already good (it may be enhanced using more points)

So, with a few tricks such as 1 and 2, and the solution of #8030 , NOODO is now sufficiently good to be exposed in the OM documentation.

@vruge do you think it is possible? I could help, adding an example derived from this toyModel, obviously amended in the mistaken parts. 
Since for the Optimica-DO there is a fully solved example in the documentation, having the same thing for NOODO would be nice.

In these examples description of a few tricks such as 1 and 2 above would be useful for the reader. Regarding #1, an indication of how to choose tolerance for the first attempts and what tolerance actually means would be useful for the readers.

**Comment by casella** [comment]
@max-privato having a short tutorial with this model of yours in the documentation would be really nice!

**Comment by casella** [comment]
@vruge it seems that you fix in #8030 had a very bad outcome on SimCode performance in some cases, see #9213.

Is there a better way to do that without ending up with O(N^2) complexity? Otherwise we need to activate what you did in #8030 conditionally when doing optimization only. For certain simulations, that is really a killer, and we have contractual obligations to improve that by end of October 😅 .

Thanks!

Adding @mahge to the loop.
